File: app/src/main/java/example/com/crackle/fragment/MovieInfoFragment.java
package example.com.crackle.fragment;


import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RatingBar;
import android.widget.TextView;
import android.widget.Toast;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import example.com.crackle.R;
import example.com.crackle.adapter.MovieVideoAdapter;
import example.com.crackle.listener.MovieApiClient;
import example.com.crackle.model.CreditResults;
import example.com.crackle.model.Crew;
import example.com.crackle.model.Movie;
import example.com.crackle.model.Video;
import example.com.crackle.utils.MovieApiService;
import example.com.crackle.utils.Utils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static example.com.crackle.utils.Constants.API_KEY;
import static example.com.crackle.utils.Constants.LINEAR_LAYOUT_HORIZONTAL;
import static example.com.crackle.utils.Constants.MOVIE;


/**
 * A simple {@link Fragment} subclass.
 */
public class MovieInfoFragment extends Fragment {

    @BindView(R.id.tmdbRating)
    TextView tmdbRating;
    @BindView(R.id.ratingBar)
    RatingBar ratingBar;
    @BindView(R.id.popularity)
    TextView popularity;
    @BindView(R.id.language)
    TextView language;
    @BindView(R.id.plot)
    TextView plotTextView;
    @BindView(R.id.director)
    TextView directorTextView;
    @BindView(R.id.release_date)
    TextView releaseDateTextView;
    @BindView(R.id.homepage)
    TextView homepage;
    @BindView(R.id.originalTitle)
    TextView originalTitle;
    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;
    @BindView(R.id.emptyTextView)
    TextView emptyTextView;


    public MovieInfoFragment() {
        // Required empty public constructor
    }

    /**
     * inflates the view for the fragment
     *
     * @param inflater           reference to inflater service
     * @param container          parent for the fragment
     * @param savedInstanceState reference to bundle object that can be used to save activity states
     * @return inflated view for fragment
     */
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_movie_info, container, false);
    }

    /**
     * called after onCreateView returns - resolve references to child views here
     *
     * @param view               reference to created view that can be modified
     * @param savedInstanceState reference to bundle object that can be used to save activity states
     */
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ButterKnife.bind(this, view);

        //initialize data set and set up the adapter
        List<Video> videoList = new ArrayList<>();
        List<Crew> crewList = new ArrayList<>();

        //set up RecyclerView - define caching properties and default animator
        Utils.setupRecyclerView(getContext(), recyclerView, LINEAR_LAYOUT_HORIZONTAL);

        //set up adapter for RecyclerView
        final MovieVideoAdapter adapter = new MovieVideoAdapter(getContext(), videoList);
        recyclerView.setAdapter(adapter);

        //fetch list of language code and corresponding names
        HashMap<String, String> languageMap = Utils.fetchAllLanguages(getContext());

        //get movie object
        if (getArguments() != null) {
            Movie movie = getArguments().getParcelable(MOVIE);
            if (movie != null) {

                //set up view data
                plotTextView.setText(movie.getPlot());
                releaseDateTextView.setText(movie.getReleaseDate());
                String rating = movie.getUserRating() == 0 ? getString(R.string.no_ratings) : DecimalFormat.getNumberInstance().format(movie.getUserRating()).concat("/10");
                tmdbRating.setText(rating);
                ratingBar.setRating((float) (movie.getUserRating() / 2f));
                popularity.setText(DecimalFormat.getNumberInstance().format(movie.getPopularity()));
                language.setText(languageMap.get(movie.getLanguage()));
                if (!TextUtils.isEmpty(movie.getHomepage())) {
                    homepage.setText(movie.getHomepage());
                }
                if (!TextUtils.isEmpty(movie.getOriginalTitle())) {
                    originalTitle.setText(movie.getOriginalTitle());
                }

                //initialize retrofit client and call object that wraps the response
                MovieApiClient client = MovieApiService.getClient().create(MovieApiClient.class);
                //invoke movie credits call passing the movie id and API KEY
                Call<CreditResults> creditResultsCall = client.getMovieCredits(movie.getMovieId(), API_KEY);
                //invoke API call asynchronously
                creditResultsCall.enqueue(new Callback<CreditResults>() {
                    @Override
                    public void onResponse(@NonNull Call<CreditResults> call, @NonNull Response<CreditResults> response) {
                        //verify if the response body or the fetched results are empty/null
                        if (response.body() == null || response.body().getCrewList() == null || response.body().getCrewList().size() == 0) {
                            return;
                        }

                        //update data set, update the views accordingly
                        crewList.addAll(response.body().getCrewList());
                        directorTextView.setText("");
                        for (int i = 0; i < crewList.size(); i++) {
                            if (crewList.get(i).getJob().equalsIgnoreCase("director")) {
                                directorTextView.append(crewList.get(i).getName());
                                if (i < crewList.size() - 1 && crewList.get(i + 1).getJob().equalsIgnoreCase("director")) {
                                    directorTextView.append("\n");
                                }
                            }
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<CreditResults> call, @NonNull Throwable t) {
                        Toast.makeText(getContext(), R.string.error_movie_director, Toast.LENGTH_SHORT).show();
                    }
                });

                //display trailer thumbnails
                if (movie.getVideoResults() != null && movie.getVideoResults().getVideos() != null && movie.getVideoResults().getVideos().size() > 0) {
                    recyclerView.setVisibility(View.VISIBLE);
                    emptyTextView.setVisibility(View.GONE);
                    videoList.addAll(movie.getVideoResults().getVideos());
                    adapter.notifyDataSetChanged();
                } else {
                    emptyTextView.setVisibility(View.VISIBLE);
                    recyclerView.setVisibility(View.GONE);
                }
            }
        }
    }


    /**
     * return new instance of fragment with movie data passed in as arguments
     *
     * @param movie reference to movie object set as one of fragment's arguments
     * @return instance of fragment
     */
    public static Fragment newInstance(Movie movie) {
        MovieInfoFragment fragment = new MovieInfoFragment();
        Bundle args = new Bundle();
        args.putParcelable(MOVIE, movie);
        fragment.setArguments(args);
        return fragment;
    }
}

File: app/src/main/java/example/com/crackle/fragment/MovieInfoFragment.kt
package example.com.crackle.fragment

import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RatingBar
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import example.com.crackle.R
import example.com.crackle.adapter.MovieVideoAdapter
import example.com.crackle.model.CreditResults
import example.com.crackle.model.Crew
import example.com.crackle.model.Movie
import example.com.crackle.model.Video
import example.com.crackle.utils.Constants
import example.com.crackle.utils.MovieApiService.client
import example.com.crackle.utils.Utils.fetchAllLanguages
import example.com.crackle.utils.Utils.setupRecyclerView
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.text.DecimalFormat

/**
 * A simple [Fragment] subclass.
 */
class MovieInfoFragment : Fragment() {
    @JvmField
    @BindView(R.id.tmdbRating)
    var tmdbRating: TextView? = null
    @JvmField
    @BindView(R.id.ratingBar)
    var ratingBar: RatingBar? = null
    @JvmField
    @BindView(R.id.popularity)
    var popularity: TextView? = null
    @JvmField
    @BindView(R.id.language)
    var language: TextView? = null
    @JvmField
    @BindView(R.id.plot)
    var plotTextView: TextView? = null
    @JvmField
    @BindView(R.id.director)
    var directorTextView: TextView? = null
    @JvmField
    @BindView(R.id.release_date)
    var releaseDateTextView: TextView? = null
    @JvmField
    @BindView(R.id.homepage)
    var homepage: TextView? = null
    @JvmField
    @BindView(R.id.originalTitle)
    var originalTitle: TextView? = null
    @JvmField
    @BindView(R.id.recyclerView)
    var recyclerView: RecyclerView? = null
    @JvmField
    @BindView(R.id.emptyTextView)
    var emptyTextView: TextView? = null

    /**
     * inflates the view for the fragment
     *
     * @param inflater           reference to inflater service
     * @param container          parent for the fragment
     * @param savedInstanceState reference to bundle object that can be used to save activity states
     * @return inflated view for fragment
     */
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? { // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_movie_info, container, false)
    }

    /**
     * called after onCreateView returns - resolve references to child views here
     *
     * @param view               reference to created view that can be modified
     * @param savedInstanceState reference to bundle object that can be used to save activity states
     */
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ButterKnife.bind(this, view)
        //initialize data set and set up the adapter
        val videoList: MutableList<Video> = ArrayList()
        val crewList: MutableList<Crew> = ArrayList()
        //set up RecyclerView - define caching properties and default animator
        setupRecyclerView(context!!, recyclerView!!, Constants.LINEAR_LAYOUT_HORIZONTAL)
        //set up adapter for RecyclerView
        val adapter = MovieVideoAdapter(context!!, videoList)
        recyclerView!!.adapter = adapter
        //fetch list of language code and corresponding names
        val languageMap = fetchAllLanguages(context!!)
        //get movie object
        if (arguments != null) {
            val movie: Movie = arguments!!.getParcelable(Constants.MOVIE)!!
            if (movie != null) { //set up view data
                plotTextView!!.text = movie.plot
                releaseDateTextView!!.text = movie.releaseDate
                val rating = if (movie.userRating == 0.0) getString(R.string.no_ratings) else DecimalFormat.getNumberInstance().format(movie.userRating) + "/10"
                tmdbRating!!.text = rating
                ratingBar!!.rating = (movie.userRating / 2f).toFloat()
                popularity!!.text = DecimalFormat.getNumberInstance().format(movie.popularity)
                language!!.text = languageMap!![movie.language]
                if (!TextUtils.isEmpty(movie.homepage)) {
                    homepage!!.text = movie.homepage
                }
                if (!TextUtils.isEmpty(movie.originalTitle)) {
                    originalTitle!!.text = movie.originalTitle
                }

                //invoke movie credits call passing the movie id and API KEY
                val creditResultsCall = client.getMovieCredits(movie.movieId, Constants.API_KEY)
                //invoke API call asynchronously
                creditResultsCall.enqueue(object : Callback<CreditResults?> {
                    override fun onResponse(call: Call<CreditResults?>, response: Response<CreditResults?>) { //verify if the response body or the fetched results are empty/null
                        if (response.body() == null || response.body()!!.crewList == null || response.body()!!.crewList.size == 0) {
                            return
                        }
                        //update data set, update the views accordingly
                        crewList.addAll(response.body()!!.crewList)
                        directorTextView!!.text = ""
                        for (i in crewList.indices) {
                            if (crewList[i].job.equals("director", ignoreCase = true)) {
                                directorTextView!!.append(crewList[i].name)
                                if (i < crewList.size - 1 && crewList[i + 1].job.equals("director", ignoreCase = true)) {
                                    directorTextView!!.append("\n")
                                }
                            }
                        }
                    }

                    override fun onFailure(call: Call<CreditResults?>, t: Throwable) {
                        Toast.makeText(context, R.string.error_movie_director, Toast.LENGTH_SHORT).show()
                    }
                })
                //display trailer thumbnails
                if (movie.videoResults != null && movie.videoResults.videos != null && movie.videoResults.videos.size > 0) {
                    recyclerView!!.visibility = View.VISIBLE
                    emptyTextView!!.visibility = View.GONE
                    videoList.addAll(movie.videoResults.videos)
                    adapter.notifyDataSetChanged()
                } else {
                    emptyTextView!!.visibility = View.VISIBLE
                    recyclerView!!.visibility = View.GONE
                }
            }
        }
    }

    companion object {
        /**
         * return new instance of fragment with movie data passed in as arguments
         *
         * @param movie reference to movie object set as one of fragment's arguments
         * @return instance of fragment
         */
        fun newInstance(movie: Movie?): Fragment {
            val fragment = MovieInfoFragment()
            val args = Bundle()
            args.putParcelable(Constants.MOVIE, movie)
            fragment.arguments = args
            return fragment
        }
    }
}